temporal-sdk/src/main/java/io/temporal/internal/worker/LocalActivityWorker.java

```diff
@@ -180,149 +243,375 @@ public void handle(LocalActivityTask task) throws Exception {
+
+        if (activityTask.hasStartToCloseTimeout()) {
+          startToCloseTimeoutFuture =
+              scheduledExecutor.schedule(
+                  new StartToCloseTimeoutHandler(attemptTask),
+                  ProtobufTimeUtils.toJavaDuration(
+                          attemptTask.getAttemptTask().getStartToCloseTimeout())
+                      .toMillis(),
+                  TimeUnit.MILLISECONDS);
+        }
+
+        metricsScope.counter(MetricsType.LOCAL_ACTIVITY_TOTAL_COUNTER).inc(1);
+
+        ActivityTaskHandler.Result activityHandlerResult;
+        Stopwatch sw = metricsScope.timer(MetricsType.LOCAL_ACTIVITY_EXECUTION_LATENCY).start();
+        try {
+          activityHandlerResult =
+              handler.handle(new ActivityTask(activityTask, () -> {}), metricsScope, true);
+        } finally {
+          sw.stop();
+        }
+
+        // Making sure that the result handling code following this statement is mutual exclusive
+        // with the start to close timeout handler.
+        boolean startToCloseTimeoutFired =
+            startToCloseTimeoutFuture != null && !startToCloseTimeoutFuture.cancel(false);
+
+        if (startToCloseTimeoutFired) {
+          // If start to close timeout fired, the result of this activity execution should be
+          // discarded.
+          // Scheduling of the next attempt is taken care by the StartToCloseTimeoutHandler.
+          return;
+        }
+
+        if (activityHandlerResult.getTaskCompleted() != null) {
+          com.uber.m3.util.Duration e2eDuration =
+              ProtobufTimeUtils.toM3DurationSinceNow(activityTask.getScheduledTime());
+          metricsScope.timer(MetricsType.LOCAL_ACTIVITY_SUCCEED_E2E_LATENCY).record(e2eDuration);
+          executionContext.callback(LocalActivityResult.completed(activityHandlerResult));
+          return;
+        }
+
+        if (activityHandlerResult.getTaskCanceled() != null) {
+          executionContext.callback(LocalActivityResult.cancelled(activityHandlerResult));
+          return;
+        }
+
+        Preconditions.checkState(
+            activityHandlerResult.getTaskFailed() != null,
+            "One of taskCompleted, taskCanceled or taskFailed must be set");
+
+        Throwable executionThrowable = activityHandlerResult.getTaskFailed().getFailure();
+        Failure executionFailure =
+            activityHandlerResult.getTaskFailed().getTaskFailedRequest().getFailure();
+
+        if (isNonRetryableApplicationFailure(activityHandlerResult)) {
+          executionContext.callback(
+              failed(
+                  activityId,
+                  RetryState.RETRY_STATE_NON_RETRYABLE_FAILURE,
+                  executionFailure,
+                  currentAttempt,
+                  null));
+          return;
+        }
+
+        if (executionThrowable instanceof Error) {
+          // TODO Error inside Local Activity shouldn't be failing the local activity call.
```

Comment:
Meant to be done in this PR?

Reply:
It's old existing behavior. It's incorrect, but it's not trivial to fix correctly. And directly relevant to the goal of this PR. I will open a separate task for it.